src/coffee_machine.rs

```diff
@@ -0,0 +1,193 @@
+//! This file contains the [`CoffeeMachine`] struct and its methods to handle [`CoffeeOrder`].
+//! Allows users to select [`CoffeeType`], [`CoffeeSize`], and customize their orders with sugar
+
+use std::{fmt::Display, io::stdin};
+use strum::IntoEnumIterator;
+
+/// the structs to represents the coffee orders
+use crate::coffee_order::{CoffeeOrder, CoffeeSize};
+use crate::coffee_types::{Affogato, CoffeeType, Espresso, IcedCoffee};
+
+/// The exit option from the menu
+const EXIT_OPTION: u32 = 0;
+
+/// The [`CoffeeMachine`] struct represents a coffee machine that allows users to order coffee.
+/// It contains a [`CoffeeOrder`] and methods to interact with the user to create a [`CoffeeOrder`].
+pub struct CoffeeMachine {
+    coffee_order: CoffeeOrder,
+}
+
+/// Implementing the Default trait for [`CoffeeMachine`] to allow easy instantiation
+impl Default for CoffeeMachine {
+    fn default() -> Self {
+        CoffeeMachine::new()
+    }
+}
+
+impl CoffeeMachine {
+    /// Creates a new instance of [`CoffeeMachine`] with a default [`CoffeeOrder`].
+    pub fn new() -> Self {
+        CoffeeMachine {
+            coffee_order: CoffeeOrder {
+                coffee_size: CoffeeSize::Small,
+                sugar_amount: 0,
+                coffee_type: CoffeeType::CoffeeAmericano,
+            },
+        }
+    }
+
+    /// Runs the [`CoffeeMachine`], allowing users to select coffee types and customize their orders.
+    pub fn run(mut self) {
+        loop {
+            // Display the menu and get the user's choice of [`CoffeeType`].
+            match self.get_order_types::<CoffeeType>() {
+                // In case of a valid choice, proceed to get the rest of the order items
+                Ok(user_choice) => match self.get_complete_order(user_choice) {
+                    Ok(_) => {
+                        println!("Your order is: {}", self.coffee_order);
+                        println!("Enjoy your coffee!");
+                    }
+                    Err(e) => {
+                        println!("{e}");
+                    }
+                },
+                // If the user chooses to exit, break the loop
+                Err(e) if e.to_string().contains("Goodbye") => {
+                    println!("{e}");
+                    break;
+                }
+                Err(e) => {
+                    println!("{e}");
+                }
+            }
+        }
+    }
+
+    /// This function gets the complete order from the user based on their choice of coffee type.
+    /// It prompts the user for additional details like espresso type, affogato type, ice cubes,
+    /// coffee size, and sugar amount, and updates the [`CoffeeOrder`] accordingly.
+    fn get_complete_order(
+        &mut self,
+        user_choice: CoffeeType,
```

Comment:
```suggestion
        user_choice: &CoffeeType,
```
In this case, it makes sense to take it as a reference, as it doesn't hurt anyone, and can make you function more generic/flexible (for the `run` it makes sense that you didn't take it as a reference, because it makes it much more clearer what the method does and how the user should run it, just once).